sys/net/gnrc/doc.txt

```diff
@@ -0,0 +1,172 @@
+ *     switch (msg.type) {
+ *         case TYPE1:
+ *             callback1();
+ *             break;
+ *         ...
+ *     }
+ * }
+ * @endcode
+ *
+ * @note As an example have a look at the event loops of @ref net_gnrc_ipv6 and @ref net_gnrc_rpl
+ *
+ * Receiving / Transmitting Packets
+ * ================================
+ *
+ * Packets can be received or transmitted by interacting with the @ref net_gnrc_netapi.
+ *
+ * Receiving Packets
+ * -----------------
+ *
+ * The reception of a @ref net_gnrc_pkt from another thread is handled by the @ref net_gnrc_netapi
+ * module. In order to receive a @ref net_gnrc_pkt of a specific type, it is necessary to register
+ * for the appropriate @ref net_gnrc_nettype first. Your thread will then be able to receive
+ * certain commands defined in the @ref net_gnrc_netapi module
+ * (e.g. @ref net_gnrc_netapi::GNRC_NETAPI_MSG_TYPE_RCV) for all @ref net_gnrc_pkt "Packets"
+ * that your thread registered for.
+ *
+ * The following example will sketch how to receive incoming and outgoing UDP traffic on port 80.
+ * @code{c}
+ * static msg_t _msg_q[Q_SZ];
+ * msg_t msg, reply;
+ * reply.type = GNRC_NETAPI_MSG_TYPE_ACK;
+ * reply.content.value = -ENOTSUP;
+ * msg_init_queue(_msg_q, Q_SZ);
+ * gnrc_pktsnip_t *pkt = NULL;
+ * gnrc_netreg_entry me_reg = { .demux_ctx = 80, .pid = thread_getpid() };
+ * gnrc_netreg_register(GNRC_NETTYPE_UDP, &me_reg);
+ *
+ * while (1) {
+ *     msg_receive(&msg);
+ *     switch (msg.type) {
+ *         case GNRC_NETAPI_MSG_TYPE_RCV:
+ *             pkt = (gnrc_pktsnip_t *) msg.content.ptr;
+ *             _handle_incoming_pkt(pkt);
+ *             break;
+ *         case GNRC_NETAPI_MSG_TYPE_SND:
+ *             pkt = (gnrc_pktsnip_t *) msg.content.ptr;
+ *             _handle_outgoing_pkt(pkt);
+ *             break;
+ *          case GNRC_NETAPI_MSG_TYPE_SET:
+ *          case GNRC_NETAPI_MSG_TYPE_GET:
+ *             msg_reply(&msg, &reply);
+ *             break;
+ *         default:
+ *             break;
+ *     }
+ * }
+ * @endcode
+ * @note It is necessary to react with a reply message to the other thread for
+ *       @ref GNRC_NETAPI_MSG_TYPE_SET and @ref GNRC_NETAPI_MSG_TYPE_GET commands
+ *
+ * Transmitting Packets
+ * --------------------
+ *
+ * To relay the @ref net_gnrc_pkt to interested threads the @ref net_gnrc_netapi offers
+ * a dispatch functionality.
+ * The following example sketches the usage and assumes a valid @ref net_gnrc_pkt named `pkt`.
+ * @code{c}
+ * if (!gnrc_netapi_dispatch_send(GNRC_NETTYPE_UDP, 80, pkt)) {
+ *      puts("Error: no thread is interested");
+ *      gnrc_pktbuf_release(pkt);
+ *      return;
+ * }
+ * @endcode
+ * The @ref net_gnrc_pkt "pkt" will be send to all threads that registered for @ref GNRC_NETTYPE_UDP
+ * and the demux context `80`. Every registered thread will receive a @ref GNRC_NETAPI_MSG_TYPE_SND
+ * command and can access the @ref net_gnrc_pkt. If @ref gnrc_netapi_dispatch_send is replaced by
```

Comment:
I prefer to append `()` to function names to make unmistakeably clear that this is a function.